lix/packages/sdk/src/file-handlers.ts

```diff
@@ -19,54 +20,58 @@ export async function handleFileInsert(args: {
 	currentAuthor?: string;
 	queueEntry: any;
 }) {
-	const pluginDiffs: {
-		diffs: DiffReport[];
-		pluginKey: string;
-	}[] = [];
-	// console.log({ args });
+	const detectedChanges: Array<DetectedChange & { pluginKey: string }> = [];
+
 	for (const plugin of args.plugins) {
 		// glob expressions are expressed relative without leading / but path has leading /
 		if (!minimatch(normalizePath(args.after.path), "/" + plugin.glob)) {
 			break;
 		}
 
-		const diffs = await plugin.diff?.file?.({
+		if (plugin.detectChanges === undefined) {
+			const error = new Error(
+				"Plugin does not support detecting changes even though the glob matches.",
+			);
+			// https://linear.app/opral/issue/LIXDK-195/await-change-queue-to-log-errors
+			console.error(error);
+			throw error;
+		}
+		for (const change of await plugin.detectChanges({
 			before: undefined,
 			after: args.after,
-		});
-
-		pluginDiffs.push({ diffs: diffs ?? [], pluginKey: plugin.key });
+		})) {
+			detectedChanges.push({
+				...change,
+				pluginKey: plugin.key,
+			});
+		}
 	}
 
 	await args.db.transaction().execute(async (trx) => {
-		for (const { diffs, pluginKey } of pluginDiffs) {
-			for (const diff of diffs ?? []) {
-				const value = diff.before ?? diff.after;
-
-				const snapshot = await trx
-					.insertInto("snapshot")
-					.values({
-						// @ts-expect-error - database expects stringified json
-						value: JSON.stringify(value),
-					})
-					.returning("id")
-					.executeTakeFirstOrThrow();
-
-				await trx
-					.insertInto("change")
-					.values({
-						type: diff.type,
-						file_id: args.after.id,
-						author: args.currentAuthor,
-						entity_id: diff.entity_id,
-						plugin_key: pluginKey,
-						snapshot_id: snapshot.id,
-						// @ts-expect-error - database expects stringified json
-						meta: JSON.stringify(diff.meta),
-						// add queueId interesting for debugging or knowning what changes were generated in same worker run
-					})
-					.execute();
-			}
+		for (const detectedChange of detectedChanges) {
+			const snapshot = await trx
+				.insertInto("snapshot")
+				.values({
+					// TODO use empty snapshot id const https://github.com/opral/lix-sdk/issues/101
+					id: detectedChange.snapshot ? undefined : "EMPTY_SNAPSHOT_ID",
+					// @ts-expect-error - database expects stringified json
```

Comment:
Assuming here that EMPTY_SNAPSHOT_ID will come with https://github.com/opral/lix-sdk/issues/101

Reply:
will be handled in follow up snapshot content addressable pr